src/ray/core_worker/reference_count.cc

```diff
@@ -778,18 +792,39 @@ void ReferenceCounter::WaitForRefRemoved(const ReferenceTable::iterator &ref_it,
                                        const rpc::WaitForRefRemovedReply &reply) {
         RAY_LOG(DEBUG) << "Received reply from borrower " << addr.ip_address << ":"
                        << addr.port << " of object " << object_id;
-        absl::MutexLock lock(&mutex_);
-
-        // Merge in any new borrowers that the previous borrower learned of.
-        const ReferenceTable new_borrower_refs =
-            ReferenceTableFromProto(reply.borrowed_refs());
-        MergeRemoteBorrowers(object_id, addr, new_borrower_refs);
+        // If the message is published, this callback will be invoked.
+        const auto message_published_callback = [this, addr,
+                                                 object_id](const rpc::PubMessage &msg) {
+          RAY_CHECK(msg.has_worker_ref_removed_message());
+          const ReferenceTable new_borrower_refs =
+              ReferenceTableFromProto(msg.worker_ref_removed_message().borrowed_refs());
+
+          absl::MutexLock lock(&mutex_);
+          CleanupBorrowersOnRefRemoved(new_borrower_refs, object_id, addr);
+          // Unsubscribe the object once the message is published.
+          RAY_CHECK(object_status_subscriber_->Unsubscribe(
+              rpc::ChannelType::WORKER_REF_REMOVED_CHANNEL, addr.ToProto(),
+              object_id.Binary()));
+        };
+
+        // If the borrower is failed, this callback will be called.
+        const auto publisher_failed_callback = [this, addr, object_id]() {
+          // When the request is failed, there's no new borrowers ref published from this
+          // borrower.
+          absl::MutexLock lock(&mutex_);
+          CleanupBorrowersOnRefRemoved({}, object_id, addr);
+        };
+
+        // If the request was failed, we just invoke the failure callback right away.
+        if (!status.ok()) {
```

Comment:
It's better to print a log here to let us have a chance to know the wrong status if it fails.
Otherwise, once it fails, we have no idea on the issue.